src/mlpack/methods/reinforcement_learning/eql_impl.hpp

```diff
@@ -0,0 +1,354 @@
+  }
+
+  arma::mat target(EnvironmentType::rewardSize, inputSize * actionSize);
+  learningNetwork.Forward(sampledStatePref, target);
+
+  const double discount = std::pow(config.Discount(), replayMethod.NSteps());
+
+  // Each slice of the cube holds the action vectors of a state-preference pair.
+  arma::cube targetCube(target.memptr(), EnvironmentType::rewardSize, actionSize,
+                        inputSize, false, true);
+  arma::cube nextActionValCube(nextActionValues.memptr(), EnvironmentType::rewardSize,
+                               actionSize, inputSize, false, true);
+
+  // Iterate over state-preference indexes (spIdx).
+  for (size_t spIdx = 0; spIdx < inputSize; ++spIdx)
+  {
+    // The multi-objective Bellman filter (H) applied over nextActionValue.
+    const arma::vec& hQ = nextActionValCube.slice(spIdx).col(bestAction(spIdx));
+
+    targetCube.slice(spIdx).col(sampledActions(spIdx).action) =
+        sampledRewardLists(spIdx) + discount * hQ * (1 - isTerminal(spIdx));
+  }
+
+  // Learn from experience.
+  arma::mat gradients;
+  learningNetwork.Backward(
+      sampledStatePref, target, extendedWeightSpace,
+      [&lambda, &extendedWeightSpace](const arma::mat& predictions,
+                                      const arma::mat& targets)
+      {
+        const size_t numElem = arma::sum((predictions - targets) != 0);
+        const double lossA =
+            std::pow(arma::norm((predictions - targets).vectorise()), 2) /
+            numElem;
+        const double lossB =
+            std::pow(arma::norm(arma::sum(extendedWeightSpace %
+                                          (predictions - targets))),
+                     2) /
+            numElem;
+
+        const double homotopyLoss = (1 - lambda) * lossA + lambda * lossB;
+
+        // Store the error.
+        arma::mat errorA = (predictions - targets) / numElem;
+        arma::mat errorB =
+            arma::sum(extendedWeightSpace % (predictions - targets)) % extendedWeightSpace;
+        const double error = 2 * ((1 - lambda) * errorA + lambda * errorB);
+        return std::make_tuple(error, homotopyLoss);
+      },
+      gradients);
+
+  replayMethod.Update(target, sampledActions, nextActionValues, gradients);
+
+  #if ENS_VERSION_MAJOR == 1
+  updater.Update(learningNetwork.Parameters(), config.StepSize(), gradients);
+  #else
+  updatePolicy->Update(learningNetwork.Parameters(), config.StepSize(),
+      gradients);
+  #endif
+
+  if (config.NoisyEQL() == true)
+  {
+    learningNetwork.ResetNoise();
+    targetNetwork.ResetNoise();
+  }
+  // Update target network.
+  if (totalSteps % config.TargetNetworkSyncInterval() == 0)
+    targetNetwork.Parameters() = learningNetwork.Parameters();
+
+  if (totalSteps > config.ExplorationSteps())
+    policy.Anneal();
+
+  if (totalSteps > config.LambdaUpdateSteps())
+  {
+    // Lambda anneal.
+    lambda += lambda_delta;
+    lambda_delta = (lambda - lambdaInit) * lambdaExpBase + lambdaInit - lambda;
```

Comment:
@zoq  Perhaps we could wrap it inside a policy? Like `lambdaUpdatePolicy`, it's a little awkward to hold state of `lambda`, `lambdaInit` & `lambda_delta` inside our `EQL` class.

Reply:
Either is fine, right now we only have a single update strategy, but I can see that a policy could make sense here anyway.